File: julia/Day18.jl
module Day18

export Submarine, on_input!, output

mutable struct Node
    val::UInt32
    directions::Vector{UInt32}
end

function depth(n::Node)
    sum(n.directions)
end

mutable struct Submarine
    all_snail_numbers::Vector{Vector{Node}}
    final_sum::Vector{Node}

    Submarine() = new([], [])
end

function on_input!(s::Submarine, a::AbstractString)
    snail_number = parse_snail_number(a)

    push!(s.all_snail_numbers, deepcopy(snail_number))
    if length(s.final_sum) == 0
        s.final_sum = snail_number
    else
        s.final_sum = add!(s.final_sum, snail_number)
    end
end

function parse_snail_number(a::AbstractString)::Vector{Node}
    result::Vector{Node} = []

    i = 1
    directions = [0, 0]

    while i <= lastindex(a)
        if a[i] == '['
            directions += [1, 0]
        elseif a[i] == ']'
            directions += [0, -1]
        elseif a[i] == ','
            directions += [-1, 1]
        else
            num_end = findnext(c -> c == ']' || c == ',', a, i) - 1
            val = parse(UInt32, a[i:num_end])

            push!(result, Node(val, copy(directions)))
            i = num_end
        end

        i += 1
    end

    result
end

function add!(s::Vector{Node}, t::Vector{Node})::Vector{Node}
    foreach(n -> n.directions += [1, 0], s)
    foreach(n -> n.directions += [0, 1], t)

    reduce!([s; t])
end

function reduce!(s::Vector{Node})::Vector{Node}
    while explode!(s) || split!(s)
    end

    s
end

function explode!(s::Vector{Node})::Bool
    for i in 1:(length(s) - 1)
        if depth(s[i]) == depth(s[i + 1]) && depth(s[i]) >= 5
            if i > 1
                s[i - 1].val += s[i].val
            end
            if i < length(s) - 1
                s[i + 2].val += s[i + 1].val
            end

            splice!(s, i:i+1, [Node(0, s[i].directions + [-1, 0])])
            return true
        end
    end

    false
end

function split!(s::Vector{Node})::Bool
    for i in 1:length(s)
        if s[i].val >= 10
            left = Node(s[i].val ÷ 2, s[i].directions + [1, 0])
            right = Node((s[i].val + 1) ÷ 2, s[i].directions + [0, 1])

            splice!(s, i:i, [left, right])
            return true
        end
    end

    return false
end

function magnitude(s::Vector{Node})::UInt64
    sum(map(n -> n.val * prod([3, 2] .^ n.directions), s))
end

function max_magnitude(as::Vector{Vector{Node}})::UInt64
    max_mag::UInt64 = 0

    for i in 1:length(as)
        for j = 1:length(as)
            if i == j
                continue
            end

            max_mag = max(max_mag, magnitude(add!(deepcopy(as[i]), deepcopy(as[j]))))
        end
    end

    max_mag
end

function output(s::Submarine)
    println("Part 1: $(magnitude(s.final_sum))")
    println("Part 2: $(max_magnitude(s.all_snail_numbers))")
end

end

File: rust/src/day18.rs
#[derive(Clone)]
struct Node {
    val: u32,
    depth: u32,
    // How many times we went to the left and to the right (to count the magnitude later).
    directions: (u32, u32),
}

pub struct Submarine {
    all_snail_numbers: Vec<Vec<Node>>,
    final_sum: Vec<Node>,
}

impl Submarine {
    pub fn new() -> Self {
        Self {
            all_snail_numbers: Vec::new(),
            final_sum: Vec::new(),
        }
    }

    pub fn on_input(&mut self, result: &str) {
        let curr_snail_number = parse_snail_number(result);
        self.all_snail_numbers.push(curr_snail_number.clone());

        if self.final_sum.len() == 0 {
            self.final_sum = curr_snail_number;
        } else {
            add(&mut self.final_sum, curr_snail_number);
            reduce(&mut self.final_sum);
        }
    }

    fn largest_magnitude(&self) -> u64 {
        let mut result = 0;

        for i in 0..self.all_snail_numbers.len() {
            for j in i + 1..self.all_snail_numbers.len() {
                let mut left = self.all_snail_numbers[i].clone();
                let right = self.all_snail_numbers[j].clone();
                add(&mut left, right);
                reduce(&mut left);
                let mag = magnitude(&left);

                result = result.max(mag);

                let left = self.all_snail_numbers[i].clone();
                let mut right = self.all_snail_numbers[j].clone();
                add(&mut right, left);
                reduce(&mut right);
                let mag = magnitude(&right);

                result = result.max(mag);
            }
        }

        result
    }

    pub fn output(&mut self) {
        println!("Part 1: {}", magnitude(&self.final_sum));
        println!("Part 2: {}", self.largest_magnitude());
    }
}

fn parse_snail_number(s: &str) -> Vec<Node> {
    let mut result = Vec::new();

    let mut depth = 0;
    let mut directions = vec![(0, 0)]; // (0, 0) is just a fake value to make directions work when parsing the first node.
    let mut i: usize = 0;

    let sb = s.as_bytes();

    let num_end_chars: &[_] = &[']', ','];

    while i < sb.len() {
        match sb[i] as char {
            '[' => {
                depth += 1;
                // Increasing depth, so we must also add to the directions vec.
                let new_direction = *directions.last().unwrap();
                directions.push((new_direction.0 + 1, new_direction.1));
            },
            ']' => {
                depth -= 1;
                // Decreasing depth, so pop from directions vec.
                directions.pop();
            },
            ',' => {
                // Finished parsing the node on the left and going to the right, so we'll change the directions.
                let last = directions.last_mut().unwrap();
                last.0 -= 1;
                last.1 += 1;
            },
            _ => { // Guaranteed to be the beginning of a number if the string is valid.
                let num_end = i + s[i..].find(num_end_chars).unwrap();
                let val = s[i..num_end].parse().unwrap();

                result.push(Node { val, depth, directions: *directions.last().unwrap() });
                // Need to subtract 1 because we'll add 1 later.
                i = num_end - 1;
            },
        }

        i += 1;
    }

    result
}

fn add(s: &mut Vec<Node>, mut t: Vec<Node>) {
    // The nodes from `s` will become the left side, so increase the left directions.
    s.iter_mut().for_each(|n| n.directions.0 += 1);
    // The nodes from `t` will become the right side, so increase the right directions.
    t.iter_mut().for_each(|n| n.directions.1 += 1);
    s.append(&mut t);
    s.iter_mut().for_each(|n| n.depth += 1);
}

fn reduce(s: &mut Vec<Node>) {
    loop {
        if explode(s) {
            continue;
        }
        if !split(s) {
            break;
        }
    }
}

fn explode(s: &mut Vec<Node>) -> bool {
    let position_to_explode = s.windows(2).position(|win| win[0].depth == win[1].depth && win[0].depth >= 5);

    if let Some(pos) = position_to_explode {
        if pos == 0 {
            s[0].val = 0;
            s[0].depth -= 1;
            // We're keeping the leftmost element of the pair, so we just need to decrease the left direction.
            s[0].directions.0 -= 1;

            let right = s.remove(1);

            // Index is 1 because we just removed the right part of the pair.
            s[1].val += right.val;
        } else if pos == s.len() - 2 {
            s[pos - 1].val += s[pos].val;
            s[pos].val = 0;
            s[pos].depth -= 1;
            // We're keeping the leftmost element of the pair, so we just need to decrease the left direction.
            s[pos].directions.0 -= 1;
            s.remove(pos + 1);
        } else {
            s[pos - 1].val += s[pos].val;
            s[pos + 2].val += s[pos + 1].val;

            // We're adding the new element on the position of the leftmost element of the pair, so we just need to decrease the left direction.
            let new_element = [Node { val: 0, depth: s[pos].depth - 1, directions: (s[pos].directions.0 - 1, s[pos].directions.1) }];
            s.splice(pos..=pos + 1, new_element);
        }

        return true;
    }

    false
}

fn split(s: &mut Vec<Node>) -> bool {
    let position_to_split = s.iter().position(|n| n.val >= 10);

    if let Some(pos) = position_to_split {
        let left = Node { val: s[pos].val / 2, depth: s[pos].depth + 1, directions: (s[pos].directions.0 + 1, s[pos].directions.1) };
        let right = Node { val: (s[pos].val + 1) / 2, depth: s[pos].depth + 1, directions: (s[pos].directions.0, s[pos].directions.1 + 1) };
        let new_elements = [left, right];

        s.splice(pos..=pos, new_elements);
        return true;
    }

    false
}

fn magnitude(s: &Vec<Node>) -> u64 {
    s.iter().map(|n| 3u64.pow(n.directions.0) * 2u64.pow(n.directions.1) * n.val as u64).sum()
}